subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/asm/ClassDependenciesVisitor.java

```diff
@@ -35,33 +38,98 @@
     private final static int API = Opcodes.ASM5;
     private static final MethodVisitor EMPTY_VISITOR = new MethodVisitor(API, null) {
     };
+    private static final Set<String> PRIMITIVES = ImmutableSet.<String>builder()
+        .add("void")
+        .add("boolean")
+        .add("char")
+        .add("short")
+        .add("int")
+        .add("long")
+        .add("float")
+        .add("double")
+        .build();
 
     private final LiteralAdapter literalAdapter;
     private final AnnotationVisitor annotationVisitor;
     private final Set<Integer> constants;
     private final Set<Integer> literals;
     private final Set<String> superTypes;
+    private final Set<String> types;
+    private final Predicate<String> typeFilter;
     private boolean isAnnotationType;
     private boolean dependencyToAll;
 
-    public ClassDependenciesVisitor(Set<Integer> constantsCollector, Set<Integer> literalsCollector) {
+    public ClassDependenciesVisitor(Set<Integer> constantsCollector) {
+        this(constantsCollector, null, null, null, null);
+    }
+
+    public ClassDependenciesVisitor(Set<Integer> constantsCollector, Set<Integer> literalsCollector, Set<String> types, Predicate<String> typeFilter, ClassReader reader) {
         super(API);
         this.constants = constantsCollector;
         this.literals = literalsCollector;
-        this.superTypes = Sets.newHashSet();
+        this.types = types;
+        this.superTypes = types == null ? null : Sets.<String>newHashSet();
         this.annotationVisitor = literals == null ? null : new LiteralRecordingAnnotationVisitor();
         this.literalAdapter = literals == null ? null : new LiteralAdapter();
+        this.typeFilter = typeFilter;
+        if (reader != null) {
+            collectClassDependencies(reader);
```

Comment:
The fact that we have some work in the constructor and some coordination code at the [usage](https://github.com/gradle/gradle/pull/1208/files#diff-ea5c619d34156ccf12fbace36a280862R49) site makes me think we should have a method that does this coordination. In the simplest case it could just be this class which has a `getClassAnalysis` method and passes itself to the reader.